### PR TITLE
ISPN-8348 Some methods from cache API do not work properly with off-heap

### PR DESCRIPTION
**Pull request by wburns**
* Fixed more bugs with encoder cache wrapper
* Document equality issue with off heap and binary

https://issues.jboss.org/browse/ISPN-8348
https://issues.jboss.org/browse/ISPN-8359
https://issues.jboss.org/browse/ISPN-8355

### CODE REVIEW

**Comment by wburns** [comment]
This should be backported to 9.1.x as well.

**Comment by gustavonalle** [comment]
LGTM, but build failed due to checkstyle errors

**Comment by wburns** [comment]
Oops fixed.

**Comment by wburns** [comment]
I have to look at some stuff still.

File: core/src/main/java/org/infinispan/commands/CommandsFactoryImpl.java [hunk 1]
```diff
@@ -200,7 +201,8 @@ public void setupDependencies(DataContainer container, CacheNotifier<Object, Obj
                                  OrderedUpdatesManager orderedUpdatesManager, StateTransferLock stateTransferLock) {
       this.dataContainer = container;
       this.notifier = notifier;
-      this.cache = cache;
+      // We shouldn't use encoded entries or anything else here
+      this.cache = AbstractDelegatingCache.unwrapCache(cache);
```
Comment (gustavonalle) [inline core/src/main/java/org/infinispan/commands/CommandsFactoryImpl.java]: FYI, ```advancedCache.withEncoding(IdentityEncoder.class)``` will avoid doing any conversion with the cache. 
Not sure if the test failures are caused by this, but the encoding stuff does not support "naked" caches atm.

**Comment by wburns** [comment]
I also found https://issues.jboss.org/browse/ISPN-8355 while fixing others. Let me update the PR and see how many failures there are now. In core there was just some around `keyFiliter` but there could be more.

**Comment by wburns** [comment]
Added in delegates for cache stream to properly encode keys provided from `CacheStream` interface. This way all encoding should be done in the `EncoderCache` for Streams and the underlying Stream just starts the stream in the uncoded state.

**Comment by wburns** [comment]
Hoping that CI passes now :)

**Comment by wburns** [comment]
My only concern is that now a stream that uses an `EncodedCache` which is all of them has 1 additional wrapper around it. If we find this is too costly we could leak the encoding code into the various `CacheStream` implementations do the unwrapping of the filtered keys as well.

**Comment by gustavonalle** [comment]
Is CI dead?

**Comment by gustavonalle** [comment]
retriggered manually

**Comment by gustavonalle** [comment]
@wburns LGTM.

WRT additional wrapping, it seems to me that an extra wrapper compared to the amount of work typically performed by a stream operation is not significant, do you have a test to run some benchmarks?

**Comment by wburns** [comment]
@gustavonalle There are no benchmark tests. But thinking more I agree, this would make inconsequential different for all DIST/REPL caches. Local/Inval cache would see the biggest difference but it should still be very negligible compared to the iteration of all entries.

**Comment by gustavonalle** [comment]
Merged in master, thanks @wburns

**Pull request merged**